File: packages/soulseek_protocol/lib/src/peer/peer_listener.dart
import 'dart:async';
import 'dart:io';
import 'dart:typed_data';

import '../connection/socket_manager.dart';
import '../messages/buffer.dart';
import '../messages/codes.dart';
import '../messages/message.dart';
import '../peer/peer_connection.dart';
import '../transfer/upload_manager.dart';

class IncomingConnection implements TransferConnection {
  final SocketManager _socketManager;
  String _username;

  IncomingConnection(this._socketManager, {String username = ''})
      : _username = username;

  @override
  String get username => _username;

  set username(String value) => _username = value;

  @override
  Stream<SoulseekMessage> get messages => _socketManager.messages;

  @override
  void sendMessage(SoulseekMessage message) => _socketManager.sendMessage(message);

  @override
  void sendRaw(int code, Uint8List payload) => _socketManager.sendRaw(code, payload);

  void dispose() => _socketManager.dispose();
}

class PeerListener {
  ServerSocket? _serverSocket;
  final List<IncomingConnection> _connections = [];
  StreamSubscription? _acceptSub;

  final UploadManager uploadManager;

  PeerListener({required this.uploadManager});

  Future<void> start(int port) async {
    _serverSocket = await ServerSocket.bind(InternetAddress.anyIPv4, port);
    _acceptSub = _serverSocket!.listen(_onConnection);
  }

  int get port => _serverSocket?.port ?? 0;
  bool get isListening => _serverSocket != null;
  int get connectionCount => _connections.length;

  void _onConnection(Socket socket) {
    final socketManager = SocketManager();
    socketManager.accept(socket);
    final connection = IncomingConnection(socketManager);
    _connections.add(connection);

    socketManager.messages.listen((message) {
      if (message.code == PeerCode.transferRequest) {
        _handleTransferRequest(message, connection);
      }
    });
  }

  void _handleTransferRequest(SoulseekMessage message, IncomingConnection connection) {
    try {
      final buffer = ReadBuffer(message.payload);
      final direction = buffer.readInt32();
      final fileCode = buffer.readInt32();
      final filename = buffer.readString();
      final fileSize = buffer.readInt32();

      final request = TransferRequest(
        direction: direction,
        fileCode: fileCode,
        filename: filename,
        fileSize: fileSize,
      );

      uploadManager.handleTransferRequest(
        request,
        connection,
      );
    } catch (_) {
      // Skip malformed transfer requests
    }
  }

  Future<void> stop() async {
    await _acceptSub?.cancel();
    _acceptSub = null;
    await _serverSocket?.close();
    _serverSocket = null;
    for (final conn in _connections) {
      conn.dispose();
    }
    _connections.clear();
  }

  void dispose() {
    stop();
  }
}

File: Soulseek.Protocol/Peer/PeerListener.cs
using System.Net;
using System.Net.Sockets;
using Soulseek.Protocol.Connection;
using Soulseek.Protocol.Messages;
using Soulseek.Protocol.Transfer;

namespace Soulseek.Protocol.Peer;

public class IncomingConnection : ITransferConnection
{
    private readonly SocketManager _socketManager;
    private string _username;

    public IncomingConnection(SocketManager socketManager, string username = "")
    {
        _socketManager = socketManager;
        _username = username;
    }

    public string Username => _username;

    public IObservable<SoulseekMessage> MessageStream => _socketManager.MessageStream;

    public void SendMessage(SoulseekMessage message) => _socketManager.SendMessage(message);
    public void SendRaw(int code, byte[] payload) => _socketManager.SendRaw(code, payload);

    public void Dispose() => _socketManager.Dispose();
}

public class PeerListener
{
    private TcpListener? _tcpListener;
    private readonly List<(IncomingConnection Connection, IDisposable Subscription)> _connections = [];
    private CancellationTokenSource? _cts;
    private Task? _listenTask;

    public UploadManager UploadManager { get; }

    public PeerListener(UploadManager uploadManager)
    {
        UploadManager = uploadManager;
    }

    public int Port { get; private set; }
    public bool IsListening => _tcpListener != null;
    public int ConnectionCount => _connections.Count;

    public Task Start(int port)
    {
        _tcpListener = new TcpListener(IPAddress.Any, port);
        _tcpListener.Start();
        Port = port;
        _cts = new CancellationTokenSource();

        _listenTask = Task.Run(() => AcceptLoop(_cts.Token));
        return Task.CompletedTask;
    }

    private async Task AcceptLoop(CancellationToken ct)
    {
        while (!ct.IsCancellationRequested)
        {
            try
            {
                var tcpClient = await _tcpListener!.AcceptTcpClientAsync(ct);
                var socketManager = new SocketManager();
                socketManager.Accept(tcpClient);
                var connection = new IncomingConnection(socketManager);
                var sub = socketManager.MessageStream.Subscribe(message =>
                {
                    if (message.Code == PeerCode.TransferRequest)
                        HandleTransferRequest(message, connection);
                });
                _connections.Add((connection, sub));
            }
            catch (OperationCanceledException) { break; }
            catch { break; }
        }
    }

    private void HandleTransferRequest(SoulseekMessage message, IncomingConnection connection)
    {
        try
        {
            var buffer = new ReadBuffer(message.Payload);
            var direction = buffer.ReadInt32();
            var fileCode = buffer.ReadInt32();
            var filename = buffer.ReadString();
            var fileSize = buffer.ReadInt32();

            var request = new TransferRequest(direction, fileCode, filename, fileSize);
            _ = UploadManager.HandleTransferRequest(request, connection);
        }
        catch
        {
            // Skip malformed transfer requests
        }
    }

    public async Task Stop()
    {
        _cts?.Cancel();
        _cts?.Dispose();
        _cts = null;

        if (_tcpListener != null)
        {
            _tcpListener.Stop();
            _tcpListener = null;
        }

        if (_listenTask != null)
        {
            try { await _listenTask; } catch { }
            _listenTask = null;
        }

        foreach (var (conn, sub) in _connections)
        {
            sub.Dispose();
            conn.Dispose();
        }
        _connections.Clear();
    }

    public void Dispose()
    {
        _ = Stop();
    }
}